examples/tcp-wait/server.rs

```diff
@@ -0,0 +1,236 @@
+
+        // Bind to local address.
+        // If error, close sockets.
+        // FIXME: https://github.com/demikernel/demikernel/issues/640
+        libos.bind(sockqd, local)?;
+
+        println!("Listening to: {:?}", local);
+
+        return Ok(Self {
+            libos,
+            sockqd,
+            nclients,
+            clients: HashSet::default(),
+            qts: Vec::default(),
+            qts_reverse: HashMap::default(),
+            clients_accepted: 0,
+            clients_closed: 0,
+        });
+    }
+
+    fn register_client(&mut self, qd: QDesc) {
+        assert_eq!(
+            self.clients.insert(qd),
+            true,
+            "client is already registered and it shouldn't be"
+        );
+    }
+
+    fn unregister_client(&mut self, qd: QDesc) {
+        assert_eq!(
+            self.clients.remove(&qd),
+            true,
+            "client isn't registered and it should be"
+        );
+    }
+
+    fn cancel_pending_operations(&mut self, qd: QDesc) -> Vec<QToken> {
+        let qts_drained: HashMap<QToken, QDesc> = self.qts_reverse.drain_filter(|_k, v| *v == qd).collect();
+        let qts_dropped: Vec<QToken> = self.qts.drain_filter(|x| qts_drained.contains_key(x)).collect();
+        qts_dropped
+    }
+
+    fn mark_completed_operation(&mut self, index: usize) -> Result<()> {
+        let qt: QToken = self.qts.remove(index);
+        self.qts_reverse
+            .remove(&qt)
+            .ok_or(anyhow::anyhow!("unregistered queue token"))?;
+        Ok(())
+    }
+
+    fn issue_accept(&mut self) -> Result<()> {
+        // If error, close sockets.
+        // FIXME: https://github.com/demikernel/demikernel/issues/640
+        let qt: QToken = self.libos.accept(self.sockqd)?;
+        self.qts_reverse.insert(qt, self.sockqd);
+        self.qts.push(qt);
+        Ok(())
+    }
+
+    fn issue_pop(&mut self, qd: QDesc) -> Result<()> {
+        // If error, close sockets.
+        // FIXME: https://github.com/demikernel/demikernel/issues/640
+        let qt: QToken = self.libos.pop(qd, None)?;
+        self.qts_reverse.insert(qt, qd);
+        self.qts.push(qt);
+        Ok(())
+    }
+
+    fn issue_close(&mut self, qd: QDesc) -> Result<()> {
+        match self.libos.close(qd) {
+            Ok(_) => Ok(()),
+            Err(e) if e.errno == libc::ECONNRESET => Ok(()),
+            Err(_) => anyhow::bail!("wait() should succeed with close()"),
+        }
+    }
+
+    fn handle_connection_establishment(&mut self, qd: QDesc) -> Result<()> {
+        self.register_client(qd);
+
+        // Pop first packet from this connection.
+        // If error, close sockets.
+        // FIXME: https://github.com/demikernel/demikernel/issues/640
+        self.issue_pop(qd)?;
+
+        self.clients_accepted += 1;
+        println!("{} clients accepted", self.clients_accepted);
+
+        Ok(())
+    }
+
+    fn handle_connection_termination(&mut self, qd: QDesc) -> Result<Vec<QToken>> {
```

Comment:
Could we rename this to `terminate_connection()`, because we are not actually reacting to a close from the remote, but we are rather doing the opposite.

Reply:
Done